File: Python/MinStack.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: MinStack.py
@Function: 最小栈
@Link: https://leetcode-cn.com/problems/min-stack/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-25
"""


import math


class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = [math.inf]

    def push(self, x: int) -> None:
        self.stack.append(x)
        self.min_stack.append(min(x, self.min_stack[-1]))

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


# --------------------------------------
if __name__ == "__main__":
    # Test on LeetCode online.
    pass

File: C++/MinStack.cpp
/**
 * @File    : MinStack.cpp
 * @Brief   : 最小栈
 * @Link    : https://leetcode-cn.com/problems/min-stack/
 * @Author  : Wei Li
 * @Date    : 2021-07-25
*/

#include <iostream>
#include <stack>

class MinStack
{
    std::stack<int> s, min_s;

public:
    MinStack() {}
    void push(int x)
    {
        s.push(x);
        if (min_s.empty() || min_s.top() >= x)
        {
            min_s.push(x);
        }
    }
    void pop()
    {
        if (!min_s.empty() && min_s.top() == s.top())
        {
            min_s.pop();
        }
        s.pop();
    }
    int top()
    {
        return s.top();
    }
    int getMin()
    {
        return min_s.top();
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
